REPL: Show input alias (e.g., \\:cat:) for Char display

ShahabSL opened the pull request:
This PR enhances the REPL display for Char types.

It adds the LaTeX/emoji input alias after the standard Unicode information, making it consistent with the help?> mode behavior.

For example, `'😼'` will now display as:
`'😼': Unicode U+1F63C (category So: Symbol, other), input as \:smirk_cat:`

**(Self-contained paragraph explaining the choice):**
Considered modifying `Base.show` directly, but opted for adding a `REPL.show_repl` method instead to avoid introducing a dependency from `Base` on the `REPL` module's symbol mapping (`symbol_latex`). This approach keeps the REPL-specific display logic contained within the REPL module.

Closes #58158

ShahabSL committed:
feat(REPL): Show input alias (e.g., \\:cat:) for Char display

mcabbott commented:
Seems like a good idea. But if I load it by `@eval REPL function show_repl(io::IO, ...`, it doesn't seem to be called -- should that work? 

I wanted to check whether `collect("αβγ")` would also take this path.

ShahabSL commented:
Thanks for the feedback! I tested the `@eval` approach on my build to understand the behavior.

After running `using REPL` first, I was able to successfully define a modified `show_repl` method using `@eval REPL function...`. When I then displayed a character like `'α'`, the output confirmed that the dynamically defined method *was* called (it printed a debug message I added and changed the output format as expected):

```julia
julia> using REPL

julia> @eval REPL function show_repl(io::IO, mime::MIME"text/plain", c::AbstractChar)
           println(stderr, ">>> Debug: Entering the @eval'd show_repl method!") # Added for testing
           show(io, mime, c)
           latex = REPL.symbol_latex(string(c))
           if !isempty(latex)
               print(io, ", === Alias from @eval'd method: ") # Changed text/color
               printstyled(io, latex, color=:yellow)
           end
       end
show_repl (generic function with 4 methods)

julia> 'α'
>>> Debug: Entering the @eval'd show_repl method!
'α': Unicode U+03B1 (category Ll: Letter, lowercase), === Alias from @eval'd method: \alpha
```

So it seems dynamic definition *does* work for this call path at runtime, although I understand rebuilding is the standard way to test source changes incorporated during the build process.

Regarding `collect("αβγ")`, you are correct that it produces a `Vector{Char}` and wouldn't hit this `AbstractChar`-specific method due to dispatch.

ararslan commented on stdlib/REPL/src/REPL.jl:
```diff
@@ -545,6 +545,17 @@ display(d::REPLDisplay, x) = display(d, MIME("text/plain"), x)
+    latex = symbol_latex(string(c))
+    if !isempty(latex)
+        print(io, ", input as ")
+        printstyled(io, latex, color=:cyan)
```
```suggestion
        printstyled(io, latex, "<tab>"; color=:cyan)
```
Including the note of the tab key brings this closer in line with how `repl_latex` displays this information.

ShahabSL replied:
Thanks! I've added the <tab> suffix to the printstyled call in the latest commit to better match the help mode display. I've also added tests covering characters with and without aliases in stdlib/REPL/test/repl.jl as requested.

ararslan commented on stdlib/REPL/src/REPL.jl:
```diff
@@ -545,6 +545,17 @@ display(d::REPLDisplay, x) = display(d, MIME("text/plain"), x)
 
 show_repl(io::IO, mime::MIME"text/plain", x) = show(io, mime, x)
 
+# new method for AbstractChar
```
```suggestion
```
It's not necessary to note that it's a new method because in the future it will no longer be new. 😄

ShahabSL replied:
Good point😄! I've removed the # new method... comment as suggested. Thanks!

ararslan reviewed (commented):
Welcome, and thanks for the contribution! This seems like a good addition to me. It would be great if you could add tests for this method that exercise the different possible code paths.

ShahabSL committed:
Feat: Display input alias for Char in REPL (#58158)

    - Adds the LaTeX/emoji input alias to the REPL display for Char.
    - Addresses feedback from PR review (include <tab>, remove comment).
    - Includes tests covering chars with/without aliases and color output.

ShahabSL commented:
Thanks so much for the welcome and the helpful feedback, @ararslan! I really appreciate you taking the time to review.

I've pushed commits addressing the suggestions:
*   Added `<tab>` to the alias output to match help mode, as suggested.
*   Removed the `# new method` comment.
*   Added tests in `stdlib/REPL/test/repl.jl` covering the different code paths (characters with and without aliases, handling color output).

The local tests for `stdlib/REPL/test/repl.jl` are passing on my end. Hopefully, this looks good now. Ready for another look when you get a chance!

ararslan commented on stdlib/REPL/test/repl.jl:
```diff
@@ -1987,3 +1987,44 @@ end
+# Issue #58158 add alias for Char display in REPL
+@testset "REPL show_repl Char alias" begin
+    # Test character with a known emoji alias
+    buf = IOBuffer()
+    io = IOContext(buf, :color => true)
+    REPL.show_repl(io, MIME("text/plain"), '😼')
+    output = String(take!(buf))
```
FYI, you can simplify this using `sprint`:
```suggestion
    output = sprint(REPL.show_repl, MIME("text/plain"), '😼'; context=(:color => true))
```

ShahabSL replied:
Thanks for the great suggestion, @ararslan! I've updated the tests in `stdlib/REPL/test/repl.jl` to use `sprint` as you recommended. It definitely makes the code cleaner.

The changes have been pushed, and the local tests are still passing. Ready for another look when you have a moment!

ShahabSL committed:
Refactor(REPL): Simplify Char display tests using sprint

This change addresses reviewer feedback on PR #58158.

The test cases for `REPL.show_repl` with `AbstractChar` in
`stdlib/REPL/test/repl.jl` have been updated to use `sprint`
for capturing output. This makes the tests more concise and
idiomatic by replacing the previous `IOBuffer` and `IOContext`
setup.

ararslan commented:
The implementation looks great to me, nice work! It occurred to me that since this is a new feature, a NEW.md entry could also be worthwhile, for example:
```diff
--- a/NEWS.md
+++ b/NEWS.md
@@ -49,6 +49,8 @@ Standard library changes

 #### REPL

+* The display of `AbstractChar`s in the main REPL mode now includes LaTeX input information like what is shown in help mode ([#58181]).
+
 #### Test

 * Test failures when using the `@test` macro now show evaluated arguments for all function calls ([#57825], [#57839]).
```

ShahabSL committed:
Docs: Add NEWS.md entry for Char display enhancement

    Documents the new feature where the REPL display of AbstractChar
    now includes LaTeX/emoji input aliases, as per reviewer suggestion
    on PR #58181.

LilithHafner committed:
Merge branch 'master' into feat/repl-emoji-alias

LilithHafner commented:
Thanks for contributing, @ShahabSL! This is a nice usability feature that some folks might come to depend on in some edge cases. Sorry it took so long to get back to you this time with another round of feedback. Everything looked good to me so I went ahead and merged it :)

ShahabSL commented:
Hi @LilithHafner, that's fantastic to hear! Thank you so much for merging this and for your kind words. I'm really glad you see it as a nice usability feature. Absolutely no worries at all about any delay, I truly appreciate you taking the time for the final review and getting it merged. It's wonderful to see it in! I'm especially thrilled as this was my first feature contribution, and it's been such a positive learning experience. 😊
